"""wf-core-guideseq

Guideseq latch pkg.

(c) 2021 by latch.ai.
"""

import os.path
import subprocess
from pathlib import Path
from typing import Annotated, List, Optional, Tuple

from flytekit import LaunchPlan, task, workflow
from flytekit.core.with_metadata import FlyteMetadata
from flytekit.types.directory import FlyteDirectory
from flytekit.types.file import FlyteFile

from latch.s3_dir_download import download_dir, ensure_dir

@task
def guideseq(
        manifest: FlyteFile,
        input_dir: FlyteDirectory,
        identify_and_filter: bool = False,
        skip_demultiplex: bool = False,
) -> FlyteDirectory:

    crispresso_cmd = [
        "CRISPResso",
        "--fastq_r1",
        fastq_r1,
        "--amplicon_seq",
        amplicon_seq,
    ] + optional_cmd_params

    # Constructing the Path objects will have the side effect of
    # pulling data from S3 if the FlyteFiles are remote
    fastq_r1_path = Path(fastq_r1)
    if fastq_r2 is not None:
        _ = Path(fastq_r2)

    if name is None or name == "":
        # Without an output_prefix specified, CRISPResso defaults to
        # using the file basename
        name = fastq_r1_path.name.split(".")[0]

    crispresso_cmd.extend(("--name", name))

    print(f"Command: {crispresso_cmd}")

    results = subprocess.run(crispresso_cmd, cwd=None, capture_output=True, text=True)

    print(f"Working directory {os.getcwd()}")
    if results is not None:
        print(f" stdout: {results.stdout}" + f" stderr: {results.stderr}")
    else:
        print("No process output generated")

    crisp_output_dir = Path(f"CRISPResso_on_{name}")
    output_html = Path(f"CRISPResso_on_{name}.html")
    if not output_html.is_file():
        raise ValueError(
            f"Expected output {output_html} was not generated by CRISPResso"
        )

    def _fmt_dir(bucket_path: str) -> str:
        if bucket_path[-1] == "/":
            return bucket_path[:-1]
        return bucket_path

    return (
        FlyteFile(
            str(output_html.resolve()),
            remote_path=_fmt_dir(output_folder.remote_source) + f"/{output_html.name}",
        ),
        FlyteDirectory(
            str(crisp_output_dir.resolve()),
            remote_directory=_fmt_dir(output_folder.remote_source)
            + f"/{crisp_output_dir.name}",
        ),
    )


@workflow
def guideseq_wf(
        manifest: FlyteFile,
        input_dir: FlyteDirectory,
        identify_and_filter: bool = False,
        skip_demultiplex: bool = False,
) -> FlyteDirectory:
    """The guideseq package implements our data preprocessing and analysis pipeline for GUIDE-Seq data. It takes a parameter manifest file (.yaml) specifying raw sequencing reads as input and produces a table of annotated off-target sites as output.

    __metadata__:
        display_name: guideseq
        author:
            name: Lihua Julie Zhu
            email: lhtsai.mit@gmail.com
            github: https://github.com/tsailabSJ
        repository: https://github.com/tsailabSJ/guideseq
        license:
            id: AGPL-3.0

    Args:
        manifest:
          File describing all inputs. See https://github.com/tsailabSJ/guideseq for instructions.
          For Latch, modify the manifest so that paths start at the input directory. For example, if you have
          an file in your input directory with path `input_dir/data/undemux.r1.fastq.gz`, write it as
          `data/undemux.r1.fastq.gz`.

          __metadata__:
            display_name: Manifest File
            appearance:
              detail: (yaml)
              placeholder: Select a file or input URL/URI...

        input_dir:
          File containing all inputs with paths matching the manifest

          __metadata__:
            display_name: Input Directory

        identify_and_filter:
          Skip the demultiplex, umitag, consolidate, align, and visualize steps. Only run identify and filter.

          __metadata__:
            display_name: Only Identify and Filter

        skip_demultiplex:
          Skip the demultiplex step. Run umitag, consolidate, align, identify, filter, and visualize

          __metadata__:
            display_name: Data is Demultiplexed
    """

    results = guideseq(
        manifest = manifest,
        input_dir = input_dir,
        identify_and_filter = identify_and_filter,
        skip_demultiplex = skip_demultiplex,
    )

    return results


LaunchPlan.create(
    "guideseq_wf.Basic",
    guideseq_wf,
    default_inputs={
    },
)

